src/map.c

```diff
@@ -3,752 +3,624 @@
-{
-    node->parent_color = (node->parent_color & ~1LU) | color;
-}
+#define RB_MAX_DEPTH (sizeof(void *) << 4)
 
-/*
- * Set parent and color of node
- * @node: pointer to the rb node
- * @parent: pointer to the new parent node
- * @color: new color of the node
- */
-static inline void rb_set_parent_color(map_node_t *node,
-                                       map_node_t *parent,
-                                       map_color_t color)
+/* Left accessors */
+static inline map_node_t *rbtnode_get_left(map_node_t *node)
 {
-    node->parent_color = (unsigned long) parent | color;
+    return node->link.left;
 }
 
-/*
- * Check if node is red
- * @node: Node to check
- *
- * Return: false when @node is NULL or not red, true if @node is red
- */
-static inline bool rb_is_red(map_node_t *node)
+static inline void rbtnode_set_left(map_node_t *node, map_node_t *left)
 {
-    return node && (rb_color(node) == RB_RED);
+    node->link.left = left;
 }
 
-/* Create a node to be attached in the map internal tree structure */
-static map_node_t *map_create_node(void *key,
-                                   void *value,
-                                   size_t ksize,
-                                   size_t vsize)
+/* Right accessors */
+static inline map_node_t *rbtnode_get_right(map_node_t *node)
 {
-    map_node_t *node = malloc(sizeof(struct map_node));
-
-    /* Allocate memory for the keys and values */
-    node->key = malloc(ksize);
-    node->data = malloc(vsize);
-
-    /* Setup the pointers */
-    node->left = node->right = NULL;
-
-    /* Set the color to read by default */
-    rb_set_parent_color(node, NULL, RB_RED);
-
-    /*
-     * Copy over the key and values
-     *
-     * If the parameter passed in is NULL, make the element blank instead of
-     * a segfault.
-     */
-    if (!key)
-        memset(node->key, 0, ksize);
-    else
-        memcpy(node->key, key, ksize);
-
-    if (!value)
-        memset(node->data, 0, vsize);
-    else
-        memcpy(node->data, value, vsize);
-
-    return node;
+    return (map_node_t *) (((uintptr_t) node->link.right_red) & ~3);
 }
 
-static void map_delete_node(map_t obj UNUSED, map_node_t *node)
+static inline void rbtnode_set_right(map_node_t *node, map_node_t *right)
 {
-    free(node->key);
-    free(node->data);
-    free(node);
+    node->link.right_red =
+        (map_node_t *) (((uintptr_t) right) |
+                        (((uintptr_t) node->link.right_red) & 1));
 }
 
-/*
- * Perform left rotation with "node". The following happens (with respect
- * to "C"):
- *
- *         B                C
- *        / \              / \
- *       A   C     =>     B   D
- *            \          /
- *             D        A
- *
- * Returns the new node pointing in the spot of the original node.
- */
-static map_node_t *map_rotate_left(map_t obj, map_node_t *node)
+/* Color accessors */
+static inline bool rbtnode_get_red(map_node_t *node)
```

Comment:
I would like to propose the following change:
1. Keep `typedef enum { RB_BLACK = 0, RB_RED } map_color_t;`
2. Reimplement `rbtnode_get_red` as following:
```c
static inline map_color_t rbnode_color(const map_node_t *node)
{
    return (bool) (((uintptr_t) node->link.right_red) & 1);
}
```
3. Replace all occurrences  `rbtnode_get_red()` with `rbnode_color() == RB_RED`

Then, the readability should be improved.